README.md
# Openfed D8

OpenFed is a general-purpose, multilingual Drupal 8 distribution.

It is developed by/for the Belgian Federal Public Service for ICT (Fedict) as
part of the Fast2Web offering.


## Usage

### Installation

The recommended way to install Openfed is using
[Composer](https://getcomposer.org/doc/00-intro.md#installation-linux-unix-osx).

You can use our [Openfed Project](https://github.com/openfed/openfed8-project/tree/10.x) template like:

```
composer create-project openfed/openfed8-project MY_PROJECT
```

After that you can just install Openfed like any other distribution.

### Update

**Important note before update**: Make sure you are running Openfed 8.x-9.11 or higher before updating to Openfed8 10.x

In order to successfully update to Openfed 8.x-10.x from a Openfed 8.x-9.x version, you should disable the following modules:
- toolbar_themes
- sharemessage
- simple_gmap
- scheduled_updates
- field_default_token
- contact_storage_clear
- yamlform_clear
- features

You should also update your theme if you are using twig_tweak 1.x, previously shipped with Openfed 8.x-9.x and earlier versions. See the following page to know what changed https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x

If these steps are not performed, Openfed be able to update.

### Server updates

Openfed 8.x-10.x requires PHP 7.3. Make sure that the right version of PHP is being used before updating.

## Bug Report

You should use, preferably, Drupal.org [Openfed issue queue](https://www.drupal.org/project/issues/openfed).


CHANGELOG.md
CHANGELOG
=========

15 March 2022 - Version 10.11
----------------------------
Improve validation script to make it compatible with older grep versions

10 March 2022 - Version 10.10
----------------------------
Improve update/upgrade script comments and made code a bit more flexible

09 February 2022 - Version 10.9
----------------------------
Improve script by doing a composer.json merge for update/upgrade
Composer.lock update

07 February 2022 - Version 10.8
----------------------------
Add Upgrade composer command
Fix local patch path

03 February 2022 - Version 10.7
----------------------------
Backport Drupal 9.x SA-CORE-2022-001

24 January 2022 - Version 10.6
----------------------------
Patch CSHS module to fix major content editing issue

07 January 2022 - Version 10.5
----------------------------
Several security updates:
- Update Linkit due to SA-CONTRIB-2021-042.
- Update Search API attachments due to SA-CONTRIB-2021-034.
- Update Super Login due to SA-CONTRIB-2022-048.

17 December 2021 - Version 10.4
----------------------------
- Revert code to fix issue with drupal-profile path

07 December 2021 - Version 10.3
----------------------------
- Update Rules version

18 November 2021 - Version 10.2
----------------------------
- Added current_page_crumb module to fix some edge cases when breadcrumb is not correctly generated
- Uninstall wbm2cm because it's unlikely that it gets ported to D9
- Reorganize libraries and add new library smillart/wai-aria-patterns-and-widgets
- Update code and ckeditor libraries due to SA-CORE-2021-011


22 September 2021 - Version 10.1
----------------------------
Fix composer.json validation condition


20 September 2021 - Version 10.0
----------------------------
This version is a transition between Drupal 8 and Drupal 9. All modules and code in this release should be compatible with Drupal 9, but it still uses Drupal 8.

Several updates have been made which may require rework on your site before update. This version may also require some changes at your server because it requires at least PHP 7.3 and Solr support is now limited. Old versions of Solr can still be used by using search_api_solr_legacy. See https://www.drupal.org/project/search_api_solr for more info.

At theme level, if you use twig_tweak module, you may have to update your theme and/or configuration to make it compatible with twig_tweak 2. For more info check https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x

Several modules were removed and should be uninstalled before update:
- toolbar_themes
- sharemessage
- simple_gmap
- scheduled_updates
- field_default_token
- contact_storage_clear
- yamlform_clear
- features

Two new scripts were added to Openfed, one to ease the update to this version from Openfed 8.x-9.x and the other to help with future updates of every project composer files when a Openfed update is needed. See the Openfed8-project README file for more information about the update script.

Almost every module was updated in this release in order to provide compatibility with Drupal 9.

**Note1**: To update to this version from an existing Openfed site, you must be sure to be running on Openfed 8.x-9.11 or higher.

**Note2**: Since toolbar_themes was removed, the administration toolbar will look different from previous versions. From now on Openfed will use the default Drupal toolbar for a better compatibility with Drupal 9.


01 June 2021 - Version 10.0-beta2
----------------------------
Update Drupal core due to Moderately critical - Cross Site Scripting - SA-CORE-2021-003


19 May 2021 - Version 10.0-beta1
----------------------------
- Several module updates.
- Remove scheduled_updates
- Remove sharemessage
- Remove simple_gmap
- Remove toolbar_themes
- Remove field_default_token
- Remove Openfed contact_storage_clear module
- Remove Openfed yamlform_clear
- Remove features --dev
- Remove drupal/drupal-extension --dev
- Set PHP version to 7.3


20 April 2021 - Version 10.0-alpha1
----------------------------
This version is intended to be fully compatible with Drupal 9. It includes:
- Several module updates.
- Remove scheduled_updates
- Remove sharemessage
- Remove simple_gmap
- Remove toolbar_themes
- Remove field_default_token
- Remove Openfed contact_storage_clear module
- Remove Openfed yamlform_clear
- Remove features --dev
- Remove drupal/drupal-extension --dev
- Set PHP version to 7.3

02 February 2021 - Version 9.3
----------------------------
- Fix Openfed SVG field translation+moderation issue

21 January 2021 - Version 9.2
----------------------------
 - Core update to version 8.9.13 due to SA-CORE-2021-001
 - d.org 3071446: Remove invalid iframe attributes from oembed
 - Make openfed_federalheader module aware of existing translations
 - Fix menu_link issues with new patches

16 December 2020 - Version 9.1
----------------------------
 - Core update to version 8.9.11
 - Issue #3188679: Add Matomo module

07 December 2020 - Version 9.0
----------------------------
 - Update Core to version 8.9.10
 - Image Effects: patch to allow smooth update
 - Update Kiso to version 2.6
 - Add image_widget_crop module
 - Add scheduler module and a deprecated status message about scheduled_updates
 - Add block_class module
 - Partial compatibility with Composer 2.0
 - Add sharemessage deprecation status message
 - Several module version updates
 - Add views_block_filter_block, memcache, geocoder, date_popup, address, optional_end_date, menu_firstchild, multiselect, conditional_fields, sitemap
 - Add update restrictions, requiring Openfed 8.x-8.7, due to taxonomy_access_fix

Issue list available at https://github.com/openfed/openfed8/milestone/6?closed=1

**Notes**:
 - Make sure you are using the latest Openfed 8.x-8.x version before updating to Openfed 8.x-9.0 (Requirement for Taxonomy Access Fix)
 - Theme must be checked due to the issue described on https://www.drupal.org/project/drupal/issues/3169918

scripts/composer/OpenfedValidations.php
<?php

namespace OpenfedProject\composer;

use Composer\Script\Event;
use Drupal\Core\DrupalKernel;
use Symfony\Component\HttpFoundation\Request;

/**
 * A class to do several validations before updating to Openfed 8.x-10.0. This
 * will help maintainers to do an informed update.
 *
 * This file is present in Openfed 8.x-9.11+ so composer can early call it in
 * Openfed 8.x-10.0 and make a check before having to update Openfed. As such,
 * this file is not supposed to be used by Openfed 8.x-9.x.
 */
class OpenfedValidations {

  /**
   *
   */
  public static function validateUpdate810(Event $event) {

    // This check will assure that other checks will be performed only if this
    // is a Drupal site, otherwise they will be ignored (like for first time
    // installations).
    if (!self::_isDrupalSite()) {
      return;
    }

    // We should run the validations only on Openfed 8.x-10.0 when we're
    // updating from Openfed 8.x-9.x.
    if (!self::_checkProjectVersion()) {
      return;
    }

    // Composer.json requires some manual updates, this will check if those
    // updates were done.
    self::_checkComposerFile();

    // Some modules were removed from Openfed 8.10 and they should be deleted
    // before updating to this version.
    self::_checkDeprecatedModules();

    // Twig Tweak module was updated so, if used, it should be checked for
    // compatibility issues.
    self::_checkTwigTweak2Compatibility();

    // self::_checkTwigTweak3Compatibility();
  }

  /**
   * Check if current site is a Drupal site.
   *
   * @return bool
   */
  private static function _isDrupalSite() {
    $output = trim(shell_exec('drush status --field="Drupal bootstrap"'));
    if (empty($output)) {
      return FALSE;
    }
    return TRUE;
  }

  /**
   * Checks if the current version is at least Openfed8 10.0.
   *
   * @return bool
   *   Return true if current version is 10.0 or more, false otherwise.
   */
  private static function _checkProjectVersion() {
    $composer_openfed = json_decode(file_get_contents('composer.openfed.json'), TRUE);
    $current_version = $composer_openfed['require']['openfed/openfed8'];
    preg_match('/(?:[\d+\.?]+[a-zA-Z0-9-]*)/', $current_version, $matches);
    // If current version is dev, we should ignore version check.
    if (strpos($current_version, 'dev') !== FALSE) {
      return FALSE;
    }

    return version_compare($matches[0], '10.0', '>=');
  }

  /**
   * Checks if composer file was updated as it should.
   *
   * @throws \ErrorException
   */
  private static function _checkComposerFile() {
    // We'll make sure that the composer merge is updated in the old composer
    // file.
    $composer_file = json_decode(file_get_contents('composer.json'), TRUE);
    $merged_composers = array_search('composer.libraries.json', $composer_file['extra']['merge-plugin']['require']);
    if (strpos($composer_file['require']['wikimedia/composer-merge-plugin'], '^1.') !== FALSE || $merged_composers !== FALSE) {
      throw new \ErrorException("Your composer.json doesn't seem to be up to date.");
    }
  }

  /**
   * Checks if deprecated modules are enabled end stops update if that's the
   * case.
   *
   * @throws \ErrorException
   */
  private static function _checkDeprecatedModules() {
    $modules_to_check = [
      'toolbar_themes',
      'sharemessage',
      'simple_gmap',
      'scheduled_updates',
      'field_default_token',
      'contact_storage_clear',
      'yamlform_clear',
      'features',
    ];
    foreach ($modules_to_check as $module) {
      $output = trim(shell_exec('drush pml --field="status" --filter="' . $module . '"'));
      if ($output == 'Enabled') {
        throw new \ErrorException("You can't proceed with Openfed update until you uninstall $module. See Openfed 8x-10.0 release notes.");
      }
    }
  }

  /**
   * Checks if Twig Tweak is enabled.
   *
   * @return bool
   */
  private static function _isTwigTweakEnabled() {
    $module = 'twig_tweak';
    $output = trim(shell_exec('drush pml --field="status" --filter="' . $module . '"'));
    if ($output == 'Enabled') {
      return TRUE;
    }
    return FALSE;
  }

  /**
   * Initiates Drupal container.
   *
   * @throws \Exception
   */
  private static function _initDrupalContainer() {
    $autoloader = require_once getcwd() . '/docroot/autoload.php';
    $request = Request::createFromGlobals();
    $kernel = DrupalKernel::createFromRequest($request, $autoloader, 'prod');
    $kernel->boot();
    $kernel->preHandle($request);
    if (PHP_SAPI !== 'cli') {
      $request->setSession($kernel->getContainer()->get('session'));
    }
  }

  /**
   * Check template for Twig Tweak 2.x compatibility issues.
   * See https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x
   *
   * @throws \ErrorException
   */
  private static function _checkTwigTweak2Compatibility() {
    if (self::_isTwigTweakEnabled()) {
      self::_initDrupalContainer();

      // 1. Check Rendering Blocks.
      $available_blocks = array_keys(\Drupal::service('plugin.manager.block')->getDefinitions());
      $search = shell_exec('find ./docroot/themes/ ./config/ -name "*.twig" | xargs grep -hi "drupal_block"');
      $pattern = '/drupal_block\([\'|"]([a-zA-Z0-9\-_]+)[\'|"]\)/';
      preg_match_all($pattern, $search, $matches);

      foreach ($matches[1] as $block_id) {
        // If one of the block matches is an available block plugin id, we
        // should not proceed.
        if (!in_array($block_id, $available_blocks)) {
          throw new \ErrorException("You should convert your theme to Twig Tweak 2.x before updating Openfed. See https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x#rendering-blocks.");
        }
      }

      // 2. Check preg_replace filter declaration.
      $search = shell_exec('find ./docroot/themes/ ./config/ -name "*.twig" | xargs grep -hi "preg_replace"');
      $pattern = '/preg_replace\([\'|"]\/?/';
      preg_match_all($pattern, $search, $matches);

      foreach ($matches[0] as $expression) {
        // If one of the block matches is an available block plugin id, we
        // should not proceed.
        if (substr($expression, -1) != '/') {
          throw new \ErrorException("You should convert your theme to Twig Tweak 2.x before updating Openfed. See https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x#preg-replace-filter.");
        }
      }

      // 3. Check Entity access.
      $twig_tweak_check_access = \Drupal::service('settings')->get('twig_tweak_check_access');
      if ($twig_tweak_check_access === FALSE) {
        throw new \ErrorException("You should convert your settings.php to Twig Tweak 2.x before updating Openfed. See https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x#entity-access.");
      }

      // 4. Check Region wrapper.
      // 5. Check Default field language.
      // 6. Check preg_replace filter declaration.
      $search = shell_exec('find ./docroot/themes/ ./config/ -name "*.twig" | xargs grep -hi "drupal_set_message"');
      $pattern = '/drupal_set_message\([\'|"]/';
      preg_match_all($pattern, $search, $matches);

      if (!empty($matches[0])) {
        throw new \ErrorException("You should convert your theme to Twig Tweak 2.x before updating Openfed. See https://www.drupal.org/docs/contributed-modules/twig-tweak/migrating-to-twig-tweak-2x#default-field-language.");
      }

    }
  }

  /**
   * Check template for Twig Tweak 3.x compatibility issues.
   * See https://git.drupalcode.org/project/twig_tweak/-/blob/3.x/docs/migration-to-3.x.md
   *
   * @throws \ErrorException
   */
  private static function _checkTwigTweak3Compatibility() {
    if (self::_isTwigTweakEnabled()) {
      // @todo for Openfed 11.
    }
  }

}
